File: PlayCover/AppInstaller/optool/operations.m
//
//  operations.m
//  optool
//  Copyright (c) 2014, Alex Zielenski
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  * Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  * Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
//  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
//  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
//  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


#import "operations.h"
#import "NSData+Reading.h"
#import "defines.h"

BOOL removeLoadEntryFromBinary(NSMutableData *binary, struct thin_header macho, NSString *payload) {
    // parse load commands to see if our load command is already there
    binary.currentOffset = macho.offset + macho.size;
    
    uint32_t num = 0;
    uint32_t cumulativeSize = 0;
    uint32_t removedOrdinal = -1;
    
    for (int i = 0; i < macho.header.ncmds; i++) {
        if (binary.currentOffset >= binary.length ||
            binary.currentOffset > macho.offset + macho.size + macho.header.sizeofcmds)
            break;
        
        uint32_t cmd  = [binary intAtOffset:binary.currentOffset];
        uint32_t size = [binary intAtOffset:binary.currentOffset + 4];
        
        // delete the bytes in all of the load commands matching the description
        switch (cmd) {
            case LC_REEXPORT_DYLIB:
            case LC_LOAD_UPWARD_DYLIB:
            case LC_LOAD_WEAK_DYLIB:
            case LC_LOAD_DYLIB: {
                
                struct dylib_command command = *(struct dylib_command *)(binary.bytes + binary.currentOffset);
                char *name = (char *)[[binary subdataWithRange:NSMakeRange(binary.currentOffset + command.dylib.name.offset, command.cmdsize - command.dylib.name.offset)] bytes];
                if ([@(name) isEqualToString:payload] && removedOrdinal == -1) {
                    LOG("removing payload from %s...", LC(cmd));
                    // remove load command
                    // remove these bytes and append zeroes to the end of the header
                    [binary replaceBytesInRange:NSMakeRange(binary.currentOffset, size) withBytes:0 length:0];
                    num++;
                    cumulativeSize += size;
                    
                    removedOrdinal = i;
                }
                
                binary.currentOffset += size;
                break;
            }
                
            //! EXPERIMENTAL: Shifting binding ordinals
            case LC_DYLD_INFO:
            case LC_DYLD_INFO_ONLY: {
                if (removedOrdinal == -1) {
                    binary.currentOffset += size;
                    break;
                }
#ifdef DEBUG
                struct dyld_info_command info;
                [binary getBytes:&info range:NSMakeRange(binary.currentOffset, size)];
                
                uint8_t *p = (uint8_t *)binary.mutableBytes + info.bind_off;
                uint32_t s = 0;
                while (s < info.bind_size) {
                    
                    uint8_t opcode = *p & BIND_OPCODE_MASK;
                    
                    p++;
                    s += (sizeof(&p));
                    
                    // we dont support opcode == BIND_OPCODE_SET_DYLIB_SPECIAL_IMM
                    if (opcode == BIND_OPCODE_SET_DYLIB_ORDINAL_ULEB) {
                        NSLog(@"ULEB");
                        
                        // Convert the ULEB into something we can use
                        uint32_t len = 0;
                        uint64_t ordinal = read_uleb128(p, &len);
                        // Decrement it if necessary
                        if (ordinal > removedOrdinal) {
                            ordinal--;
                            
                            // Write it back if necessary
                            write_uleb128(p, ordinal, len);
                        }
                        
                        // increment s by the size of the ULEB
                        s += len;

                        
                    } else if (opcode == BIND_OPCODE_SET_DYLIB_ORDINAL_IMM) {
                        uint8_t immediate = *p & BIND_IMMEDIATE_MASK;
                        
                        // decrement this immediate if this opcode is greater than the one we removed
                        if (immediate > i) {
                            immediate--;
                            
                            // no need to shift the opcode since we never shifted it right
                            uint8_t binding = opcode | (immediate & BIND_IMMEDIATE_MASK);
                            // reset the binding ordinal
                            *p = binding;
                        }
                    }
                }
#endif
                binary.currentOffset += size;
                break;
            }
            default:
                binary.currentOffset += size;
                break;
        }
    }
    
    if (num == 0)
        return NO;
    
    // fix the header
    macho.header.ncmds -= num;
    macho.header.sizeofcmds -= cumulativeSize;
    
    unsigned int zeroByte = 0;
    
    // append a null byte for every one we removed to the end of the header
    [binary replaceBytesInRange:NSMakeRange(macho.offset + macho.header.sizeofcmds + macho.size, 0) withBytes:&zeroByte length:cumulativeSize];
    [binary replaceBytesInRange:NSMakeRange(macho.offset, sizeof(macho.header))
                      withBytes:&macho.header
                         length:sizeof(macho.header)];
    
    return YES;
}

BOOL binaryHasLoadCommandForDylib(NSMutableData *binary, NSString *dylib, uint32_t *lastOffset, struct thin_header macho) {
    binary.currentOffset = macho.size + macho.offset;
    unsigned int loadOffset = (unsigned int)binary.currentOffset;

    // Loop through compatible LC_LOAD commands until we find one which points
    // to the given dylib and tell the caller where it is and if it exists
    for (int i = 0; i < macho.header.ncmds; i++) {
        if (binary.currentOffset >= binary.length ||
            binary.currentOffset > macho.offset + macho.size + macho.header.sizeofcmds)
            break;
        
        uint32_t cmd  = [binary intAtOffset:binary.currentOffset];
        uint32_t size = [binary intAtOffset:binary.currentOffset + 4];
        
        switch (cmd) {
            case LC_REEXPORT_DYLIB:
            case LC_LOAD_UPWARD_DYLIB:
            case LC_LOAD_WEAK_DYLIB:
            case LC_LOAD_DYLIB: {
                struct dylib_command command = *(struct dylib_command *)(binary.bytes + binary.currentOffset);
                char *name = (char *)[[binary subdataWithRange:NSMakeRange(binary.currentOffset + command.dylib.name.offset, command.cmdsize - command.dylib.name.offset)] bytes];
                
                if ([@(name) isEqualToString:dylib]) {
                    *lastOffset = (unsigned int)binary.currentOffset;
                    return YES;
                }
                
                binary.currentOffset += size;
                loadOffset = (unsigned int)binary.currentOffset;
                break;
            }
            default:
                binary.currentOffset += size;
                break;
        }
    }
    
    if (lastOffset != NULL)
        *lastOffset = loadOffset;
    
    return NO;
}

BOOL insertLoadEntryIntoBinary(NSString *dylibPath, NSMutableData *binary, struct thin_header macho) {
    uint32_t type = LC_LOAD_DYLIB;

    // parse load commands to see if our load command is already there
    uint32_t lastOffset = 0;
    if (binaryHasLoadCommandForDylib(binary, dylibPath, &lastOffset, macho)) {
        // there already exists a load command for this payload so change the command type
        uint32_t originalType = *(uint32_t *)(binary.bytes + lastOffset);
        if (originalType != type) {
            LOG("A load command already exists for %s. Changing command type from %s to desired %s", dylibPath.UTF8String, LC(originalType), LC(type));
            [binary replaceBytesInRange:NSMakeRange(lastOffset, sizeof(type)) withBytes:&type];
        } else {
            LOG("Load command already exists");
        }
        
        return YES;
    }
    
    // create a new load command
    unsigned int length = (unsigned int)sizeof(struct dylib_command) + (unsigned int)dylibPath.length;
    unsigned int padding = (8 - (length % 8));
    
    // check if data we are replacing is null
    NSData *occupant = [binary subdataWithRange:NSMakeRange(macho.header.sizeofcmds + macho.offset + macho.size,
                                                            length + padding)];

    // All operations in optool try to maintain a constant byte size of the executable
    // so we don't want to append new bytes to the binary (that would break the executable
    // since everything is offset-based–we'd have to go in and adjust every offset)
    // So instead take advantage of the huge amount of padding after the load commands
    if (strcmp([occupant bytes], "\0")) {
        NSLog(@"cannot inject payload into %s because there is no room", dylibPath.fileSystemRepresentation);
        return NO;
    }
    
    LOG("Inserting a %s command for architecture: %s", LC(type), CPU(macho.header.cputype));
    
    struct dylib_command command;
    struct dylib dylib;
    dylib.name.offset = sizeof(struct dylib_command);
    dylib.timestamp = 2; // load commands I've seen use 2 for some reason
    dylib.current_version = 0;
    dylib.compatibility_version = 0;
    command.cmd = type;
    command.dylib = dylib;
    command.cmdsize = length + padding;
    
    unsigned int zeroByte = 0;
    NSMutableData *commandData = [NSMutableData data];
    [commandData appendBytes:&command length:sizeof(struct dylib_command)];
    [commandData appendData:[dylibPath dataUsingEncoding:NSASCIIStringEncoding]];
    [commandData appendBytes:&zeroByte length:padding];
    
    // remove enough null bytes to account of our inserted data
    [binary replaceBytesInRange:NSMakeRange(macho.offset + macho.header.sizeofcmds + macho.size, commandData.length)
                      withBytes:0
                         length:0];
    // insert the data
    [binary replaceBytesInRange:NSMakeRange(lastOffset, 0) withBytes:commandData.bytes length:commandData.length];
    
    // fix the existing header
    macho.header.ncmds += 1;
    macho.header.sizeofcmds += command.cmdsize;
    
    // this is safe to do in 32bit because the 4 bytes after the header are still being put back
    [binary replaceBytesInRange:NSMakeRange(macho.offset, sizeof(macho.header)) withBytes:&macho.header];
    
    return YES;
}

File: PlayCover/AppInstaller/optool/Operations.swift
//
//  Operations.swift
//  PlayCover
//
//  Created by Isaac Marovitz on 20/09/2022.
//

import Foundation

class Operations {
    @discardableResult
    public static func removeLoadEntryFromBinary(_ binary: inout OffsetNSData,
                                                 _ inputMacho: thin_header,
                                                 _ payload: String) -> Bool {
        var macho = inputMacho
        binary.offset = Int(macho.offset + macho.size)

        var num: UInt32 = 0
        var cumulativeSize: UInt32 = 0
        var removedOrdinal: UInt32 = UInt32.max

        for index in 0..<macho.header.ncmds {
            if binary.offset >= binary.data.length
                || binary.offset > macho.offset + macho.size + macho.header.sizeofcmds {
                break
            }

            let cmd: UInt32 = binary.intAtOffset(offset: binary.offset)
            let size: UInt32 = binary.intAtOffset(offset: binary.offset + 4)

            switch cmd {
            case LC_REEXPORT_DYLIB,
            LC_LOAD_UPWARD_DYLIB,
            LC_LOAD_WEAK_DYLIB,
            UInt32(LC_LOAD_DYLIB):
                let command: dylib_command = (binary.data.bytes + binary.offset).load(as: dylib_command.self)
                let name = binary.data.subdata(with:
                                                NSRange(location: binary.offset + Int(command.dylib.name.offset),
                                                        length: Int(command.cmdsize - command.dylib.name.offset)))
                    .withUnsafeBytes({ $0.load(as: UnsafeMutablePointer<CChar>.self) })
                if NSNumber(pointer: name).isEqual(to: payload) && removedOrdinal == UInt32.max {
                    print("Removing payload from \(Patch.loadCommand(cmd))...")
                    binary.data.replaceBytes(in: NSRange(location: binary.offset,
                                                         length: Int(size)), withBytes: nil, length: 0)
                    num += 1
                    cumulativeSize += size

                    removedOrdinal = index
                }

                binary.offset += Int(size)
            default:
                binary.offset += Int(size)
            }
        }

        if num == 0 {
            return false
        }

        macho.header.ncmds -= num
        macho.header.sizeofcmds -= cumulativeSize

        var zeroByte: UInt = 0
        binary.data.replaceBytes(in: NSRange(location: Int(macho.offset + macho.header.sizeofcmds + macho.size),
                                             length: 0), withBytes: &zeroByte,
                                 length: Int(cumulativeSize))
        binary.data.replaceBytes(in: NSRange(location: Int(macho.offset),
                                             length: MemoryLayout.size(ofValue: macho.header)),
                                 withBytes: &macho.header,
                                 length: MemoryLayout.size(ofValue: macho.header))

        return true
    }

    public static func binaryHasLoadCommandForDylib(binary: inout OffsetNSData,
                                                    dylib: String,
                                                    lastOffseet: inout UInt32?,
                                                    macho: thin_header) -> Bool {
        binary.offset = Int(macho.size + macho.offset)
        var loadOffset: UInt = UInt(binary.offset)

        for _ in 0..<macho.header.ncmds {
            if binary.offset >= binary.data.length
                || binary.offset > macho.offset + macho.size + macho.header.sizeofcmds {
                break
            }

            let cmd: UInt32 = binary.intAtOffset(offset: binary.offset)
            let size: UInt32 = binary.intAtOffset(offset: binary.offset + 4)

            switch cmd {
            case LC_REEXPORT_DYLIB,
            LC_LOAD_UPWARD_DYLIB,
            LC_LOAD_WEAK_DYLIB,
            UInt32(LC_LOAD_DYLIB):
                let command: dylib_command = (binary.data.bytes + binary
                    .offset).load(as: dylib_command.self)

                let name = binary.data.subdata(with:
                                                NSRange(location: binary.offset + Int(command.dylib.name.offset),
                                                        length: Int(command.cmdsize - command.dylib.name.offset)))
                    .withUnsafeBytes({ $0.load(as: UnsafeMutablePointer<CChar>.self) })
                if NSNumber(pointer: name).isEqual(to: dylib) {
                    lastOffseet = UInt32(binary.offset)
                    return true
                }

                binary.offset += Int(size)
                loadOffset = UInt(binary.offset)
            default:
                binary.offset += Int(size)
            }
        }

        if lastOffseet != nil {
            lastOffseet = UInt32(loadOffset)
        }

        return false
    }

    public static func insetLoadEntryIntoBinary(_ dylibPath: String,
                                                _ binary: inout OffsetNSData,
                                                _ inputMacho: thin_header) -> Bool {
        var macho = inputMacho
        var type: UInt32 = UInt32(LC_LOAD_DYLIB)

        // TODO: FIGURE OUT THIS VALUE
        var lastOffset: UInt32?
        if binaryHasLoadCommandForDylib(binary: &binary, dylib: dylibPath, lastOffseet: &lastOffset, macho: macho) {
            let originalType = (binary.data.bytes + Int(lastOffset ?? UInt32(binary.offset))).load(as: UInt32.self)
            if originalType != type {
                print("A load commmand already exists for \(dylibPath). Changing command type from \(Patch.loadCommand(originalType)) to desired \(Patch.loadCommand(type))")
                binary.data.replaceBytes(in: NSRange(location: Int(lastOffset ?? UInt32(binary.offset)), length: MemoryLayout.size(ofValue: type)), withBytes: &type)
            } else {
                print("Load command already exists")
            }

            return true
        }

        let length = UInt(MemoryLayout.stride(ofValue: dylib_command.self) + dylibPath.lengthOfBytes(using: .utf8))
        let padding = UInt(8 - (length % 8))

        let occupant: NSData = binary.data.subdata(with:
                                                    NSRange(location: Int(macho.header.sizeofcmds + macho.offset + macho.size),
                                                                 length: Int(length + padding))) as NSData

        if strcmp(occupant.bytes, "\0") != 0 {
            print("Cannot inject payload into \(dylibPath) because there is no room")
            return false
        }

        print("Inserting a \(Patch.loadCommand(type)) command for architecture: \(Patch.cpu(macho.header.cputype))")

        var command: dylib_command = dylib_command()
        var dylib: dylib = dylib()
        dylib.name.offset = UInt32(MemoryLayout.stride(ofValue: dylib_command.self))
        dylib.timestamp = 2
        dylib.current_version = 0
        dylib.compatibility_version = 0
        command.cmd = type
        command.dylib = dylib
        command.cmdsize = UInt32(length + padding)

        var zeroByte: UInt = 0
        let commandData: NSMutableData = NSMutableData()
        commandData.append(&command, length: MemoryLayout.stride(ofValue: dylib_command.self))
        commandData.append(dylibPath.data(using: .ascii)!)
        commandData.append(&zeroByte, length: Int(padding))

        binary.data.replaceBytes(in: NSRange(location: Int(macho.offset + macho.header.sizeofcmds + macho.size),
                                            length: commandData.length),
                                 withBytes: nil,
                                 length: 0)

        binary.data.replaceBytes(in: NSRange(location: Int(lastOffset ?? UInt32(binary.offset)),
                                            length: 0),
                                 withBytes: commandData.bytes,
                                 length: commandData.length)

        macho.header.ncmds += 1
        macho.header.sizeofcmds += command.cmdsize

        binary.data.replaceBytes(in: NSRange(location: Int(macho.offset),
                                             length: MemoryLayout.size(ofValue: macho.header)),
                                 withBytes: &macho.header)

        return true
    }
}
